graphql/database/base/RedisModel.js

```diff
@@ -0,0 +1,107 @@
+import Redis from 'ioredis'
+import Model from './Model'
+import {
+  DatabaseConditionalCheckFailedException,
+  DatabaseItemDoesNotExistException,
+  FieldDoesNotExistException,
+  UnauthorizedQueryException,
+} from '../../utils/exceptions'
+
+const client = new Redis(process.env.UPSTASH_HOST)
+
+class RedisModel extends Model {
+  static async getInternal(keys) {
+    const key = this.getRedisKey(keys)
+    const item = await client.hgetall(key)
+    if (Object.keys(item).length === 0) {
+      throw new DatabaseItemDoesNotExistException()
+    }
+    return this.wrapItem(item)
+  }
+
+  static async getBatchInternal(keys) {
+    const items = keys.map(key => this.getInternal(key))
+    return Promise.all(items)
+  }
+
+  static async createInternal(item, overwrite = false) {
+    const redisKey = this.getRedisKey(item.id)
+    if (!overwrite) {
+      const existingEntry = await client.hgetall(redisKey)
+      if (Object.keys(existingEntry).length !== 0) {
+        throw new DatabaseConditionalCheckFailedException()
+      }
+    }
+
+    await Promise.all(
+      Object.entries(item).map(([key, value]) =>
+        client.hset(redisKey, key, value)
+      )
+    )
+    return this.wrapItem(client.hgetall(redisKey))
+  }
+
+  static async updateInternal(item) {
+    const redisKey = this.getRedisKey(item.id)
+    const existingEntry = await client.hgetall(redisKey)
+    if (Object.keys(existingEntry).length === 0) {
+      throw new DatabaseConditionalCheckFailedException()
+    }
+
+    Object.entries(item).forEach(([key, value]) => {
+      client.hset(redisKey, key, value)
+    })
+
+    return this.wrapItem(client.hgetall(redisKey))
+  }
+
+  static async updateField(userContext, id, field, value) {
+    if (!this.isQueryAuthorized(userContext, 'update', id)) {
+      throw new UnauthorizedQueryException()
+    }
+
+    if (this.schema[field] === undefined) {
+      throw new FieldDoesNotExistException()
+    }
+
+    const redisKey = this.getRedisKey(id)
+    const result = await client.hget(redisKey, field)
+
+    if (result === null) {
+      throw new DatabaseItemDoesNotExistException()
+    }
+
+    await client.hset(redisKey, field, value)
+    return client.hget(redisKey, field)
+  }
+
+  static async getField(userContext, id, field) {
+    if (!this.isQueryAuthorized(userContext, 'get', id)) {
+      throw new UnauthorizedQueryException()
+    }
+
+    if (this.schema[field] === undefined) {
+      throw new FieldDoesNotExistException()
+    }
+
+    const redisKey = this.getRedisKey(id)
+    const result = await client.hget(redisKey, field)
+    if (result === null) {
+      throw new DatabaseItemDoesNotExistException()
+    }
+
+    return result
+  }
+
+  static async wrapItem(dataPromise) {
+    return {
+      attrs: await dataPromise,
+    }
+  }
+
+  static getRedisKey(hashKey) {
+    return `${this.name}_${hashKey}`
```

Comment:
I think for the sake of simplicity we can just use a singular key for the model. I'm not sure what value a range key gets us since querying Redis for a range of keys is a super expensive operation.

Reply:
Do you see an advantage for supporting a complex key over mostly treating the RedisModel closer to key/value store?

Reply:
@jedtan I'm sorry, I apparently missed this notification. A composite key could be useful if we need to find an item based only on two other known attributes, such as "look up a user's impact from their user ID and a cause ID". With the current implementation, we're limited to creating unique identifiers and storing a reference to those IDs elsewhere. That works for now, and composite key behavior wouldn't be too hard to add if needed.